### ISSUE TITLE
Using FastText subword embeddings with Gluon

### ISSUE THREAD

**Issue by charlieyou** [dmlc/gluon-nlp#879]
I am trying to understand how to use FastText embeddings in a Gluon model. I've been looking at [this section of the documentation](https://gluon-nlp.mxnet.io/examples/word_embedding/word_embedding.html#Using-Pre-trained-Word-Embeddings-in-Gluon), but it does not address how to use the `gluon.nn.Embedding` layer when you have out-of-vocab words and want to take advantage of FTs subword functionality.

I do realize that I could grab the embeddings first and then feed that as input data to the model, but I would like to be able to fine-tune the embeddings during the training process as well.

Thank you for your help!

**Comment by leezu** [comment]
You can use the respective Block: https://gluon-nlp.mxnet.io/api/modules/model.train.html#gluonnlp.model.train.FasttextEmbeddingModel
Note that this Block takes a "Sparse array containing weights for every word and subword index." as an input. So if a your input word has index i and subword indices x, y, z, the corresponding row of the CSRNDArray should be 0 everywhere but at index i, x, y and z. At those fore indices it should be set to 0.25.

If you want to load some pretrained fasttext embeddings available via https://gluon-nlp.mxnet.io/api/modules/embedding.html#gluonnlp.embedding.create you should set `load_ngrams=True` and the `FasttextEmbeddingModel` Block will be exposed as the `unknown_lookup` attribute of the resulting object.

Please let me know if you have further questions.

**Comment by leezu** [comment]
https://github.com/dmlc/gluon-nlp/blob/5d25058a2b61fb06f6a18e39608d15a51db144d3/scripts/word_embeddings/data.py#L376-L427 provides an example how to construct the array.

**Comment by charlieyou** [comment]
Thank you for the response!

When calling `FasttestEmbeddingModel.hybrid_forward` via `__call__`, what should the `weight` parameter be? `FasttestEmbeddingModel.weight`?

**Comment by leezu** [comment]
Hi @charlieyou, the `weight` parameter will be automatically passed by the `HybridBlock`. You can just ignore it. It's an artifact of how HybridBlock currently distinguishes between the `mx.sym` and `mx.nd` API (which we plan to clean up in the coming months). `weight` will be (automatically) passed as `FasttestEmbeddingModel.weight.data(ctx)` or `FasttestEmbeddingModel.weight.var()` depending on if the Block is hybridized or not.

**Comment by charlieyou** [comment]
Great thanks! I will try this and get back to you if I have any issues

**Comment by leezu** [comment]
Hi @charlieyou, indeed you need to "flatten" to 2D and reshape back to 3D. For efficiency, you likely want to pad each training sample in a batch to have the same number of embeddings. For the "pad" embeddings, the CSRNDArray can just contain a completely empty row.

If you feel this work could be supported by some GluonNLP API, you would be very welcome to contribute such helper function.

Btw, instead of the SkipGram training code snippet linked above (https://github.com/dmlc/gluon-nlp/issues/879#issuecomment-522349864), this one is more appropriate:
https://github.com/dmlc/gluon-nlp/blob/5d25058a2b61fb06f6a18e39608d15a51db144d3/src/gluonnlp/model/train/embedding.py#L375-L422
